### PR TITLE
Switch to java 8

### PR DESCRIPTION
**Pull request by jkwatson**
And Android level 24

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -148,8 +146,8 @@ configure(opentelemetryProjects) {
         mavenLocal()
     }
 
-    sourceCompatibility = 1.7
-    targetCompatibility = 1.7
+    sourceCompatibility = 1.8
```
Comment (iNikem) [inline build.gradle:149]: Quick note: https://docs.gradle.org/current/userguide/building_java_projects.html#sec:java_cross_compilation

Comment (jkwatson) [reply build.gradle:149]: we can only use the new flag if we're compiling with java9+, though, correct?

**Review by bogdandrutu — APPROVED** [review]

File: build.gradle (hunk 1 above)
Comment (jkwatson) [reply build.gradle:149]: I tried switching to the new flag and it build locally with java 14, but failed to build on CI.

**Review by trask — APPROVED** [review]

File: api/build.gradle [hunk 2]
```diff
@@ -14,8 +14,8 @@ dependencies {
 
     annotationProcessor libraries.auto_value
 
-    signature "org.codehaus.mojo.signature:java17:1.0@signature"
-    signature "net.sf.androidscents.signature:android-api-level-14:4.0_r4@signature"
+    signature "org.codehaus.mojo.signature:java18:1.0@signature"
+    signature "net.sf.androidscents.signature:android-api-level-24:7.0_r2@signature"
```
Comment (anuraaga) [inline api/build.gradle:18]: Okhttp uses API 21 signature and once before we had discussed following what they do to get Java 8 support in a way that follows the most used Android library. Is it possible to stick with that?

https://github.com/square/okhttp/blob/master/build.gradle#L153

Though if we just avoid `CompletableFuture` I think any signature there should be fine too.

File: build.gradle (hunk 1 above)
Comment (anuraaga) [reply build.gradle:149]: Yeah we'll want to switch to using Java 11 to build on CI and drop support for building on older to take advantage of the new features. Even 15 would be ok there are usually new compilation features (often javadoc improvements) and still works on older. We would need to change the CI to have a build and separate test jdk configured with Gradle to continue testing with Java 8.

For another issue of course :)

**Review by anuraaga — APPROVED** [review]

File: api/build.gradle (hunk 2 above)
Comment (jkwatson) [reply api/build.gradle:18]: I don't have a strong opinion on this one. If `CompletableFuture` is the only thing we lose between 21 and 24 then I'm cool with that.

**Review by anuraaga — COMMENTED** [review]
We can revisit the Android API level later let's go ahead and merge first!

**Pull request merged**